src/latest/dbt_yml_files.py

```diff
@@ -0,0 +1,1009 @@
+    SimpleMetric, DerivedMetric, CumulativeMetric, ConversionMetric, RatioMetric
+]
+
+
+class QueryParams(BaseModel):
+    dimensions: Optional[ArrayOfStrings] = ""
+    metrics: Optional[ArrayOfStrings] = ""
+    where: Optional[ArrayOfStrings] = ""
+
+
+class SavedQuery(BaseModel):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    name: str
+    description: str
+    exports: Optional[List[Export]] = ""
+    label: str
+    query_params: QueryParams
+
+
+class UnitTestConfig(BaseModel):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    meta: Optional[Dict[str, Any]] = ""
+    tags: Optional[StringOrArrayOfStrings] = ""
+
+
+class UnitTest(BaseModel):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    name: str
+    description: Optional[str] = ""
+    config: Optional[UnitTestConfig] = Field(None, title="Unit Test Config")
+    expect: Expect
+    given: Optional[List[Given]] = [UnitTestGivenDict(input="", rows=[{"col1": "value1", "col2": "value2"}])]
+    model: str = Field(
+        ...,
+        description="The name of the model whose behaviour you are testing. Does not need to be wrapped in a ref.",
+        examples=["my_model"],
+    )
+    overrides: Optional[Overrides] = ""
+
+
+class BooleanOrJinjaString(RootModel[Union[JinjaString, bool]]):
+    root: Union[JinjaString, bool]
+
+
+class Constraint(BaseModel):
+    name: Optional[str] = ""
+    type: str
+    columns: Optional[StringOrArrayOfStrings] = ""
+    expression: Optional[str] = ""
+    warn_unenforced: Optional[BooleanOrJinjaString] = ""
+    warn_unsupported: Optional[BooleanOrJinjaString] = ""
+
+
+class Constraints(RootModel[List[Constraint]]):
+    root: List[Constraint]
+
+
+class ConversionTypeParams(BaseModel):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    base_measure: MetricInputMeasure
+    calculation: Optional[Calculation] = Calculation.conversion_rate
+    constant_properties: Optional[List[ConstantProperty]] = ""
+    conversion_measure: MetricInputMeasure
+    entity: str = Field(..., description="The entity to calculate over")
+    window: Optional[str] = ""
+
+
+class ConversionMetricTypeParams(BaseModel):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    conversion_type_params: Optional[ConversionTypeParams] = ""
+
+
+class CumulativeMetricTypeParams(BaseModel):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    grain_to_date: Optional[TimeGranularity] = TimeGranularity.month
+    measure: Optional[MetricInputMeasure] = ""
+    window: Optional[str] = ""
+
+
+class DerivedMetricTypeParams(BaseModel):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    expr: Optional[str] = ""
+    metrics: Optional[List[MetricInputSchema]] = ""
+
+
+class DimensionTypeParams(BaseModel):
+    model_config = ConfigDict(
+        extra="forbid",
+    )
+    time_granularity: TimeGranularity
+    validity_params: Optional[ValidityParams] = ""
```

Comment:
This typing doesn't seem correct? Empty string is not None | ValidityParams

Reply:
Yeah this is to prevent the autocomplete from outputting null. An alternative is to instead specify the default in the json_schema_extra of the `Field`, which would mean this default could stay as None. Is the difference semantically important to Instructor?

Reply:
I don't know if there are semantic differences, but it causes typing errors:
![Screenshot 2024-09-06 at 10 04 05 AM](https://github.com/user-attachments/assets/348f464c-0949-4073-8a17-377ac23f08b2)

Reply:
Ahh got it, I don't have whatever extension that is so vs code didn't flag it to me. In that case if we keep this we'll have to put everything in the json_schema_extra space instead which is all good